src/main/java/org/truffleruby/core/array/ArrayEachIteratorNode.java

```diff
@@ -42,36 +42,17 @@ public static ArrayEachIteratorNode create() {
     public abstract RubyArray execute(RubyArray array, RubyProc block, int startAt,
             ArrayElementConsumerNode consumerNode);
 
-    @Specialization(
-            guards = { "array.size == 1", "startAt == 0" },
-            limit = "storageStrategyLimit()")
-    protected RubyArray iterateOne(RubyArray array, RubyProc block, int startAt, ArrayElementConsumerNode consumerNode,
-            @Bind("array.store") Object store,
-            @CachedLibrary("store") ArrayStoreLibrary stores) {
-
-        consumerNode.accept(array, block, stores.read(store, 0), 0);
-
-        if (array.size > 1) {
-            // Implicitly profiles through lazy node creation
-            return getRecurseNode().execute(array, block, 1, consumerNode);
-        }
-
-        return array;
-    }
-
-    @Specialization(
-            guards = { "array.size != 1" },
-            limit = "storageStrategyLimit()")
+    @Specialization(limit = "storageStrategyLimit()")
     protected RubyArray iterateMany(RubyArray array, RubyProc block, int startAt, ArrayElementConsumerNode consumerNode,
-            @Bind("array.store") Object store,
-            @CachedLibrary("store") ArrayStoreLibrary stores,
+            @CachedLibrary("array.store") ArrayStoreLibrary stores,
             @Cached LoopConditionProfile loopProfile,
+            @Cached("createIdentityProfile()") IntValueProfile arraySizeProfile,
             @Cached ConditionProfile strategyMatchProfile) {
         int i = startAt;
         try {
-            for (; loopProfile.inject(i < array.size); i++) {
+            for (; loopProfile.inject(i < arraySizeProfile.profile(array.size)); i++) {
                 if (strategyMatchProfile.profile(stores.accepts(array.store))) {
-                    consumerNode.accept(array, block, stores.read(store, i), i);
+                    consumerNode.accept(array, block, stores.read(array.store, i), i);
```

Comment:
Could you save array.store in a local variable just above this line?
So at least the store we accepted (stores.accepts()) is the store we read from (stores.read()).